rfcs/0042-networking-refactor.md

```diff
@@ -0,0 +1,334 @@
+      ; txn_pool_diffs: Transaction_pool.Diff.t Strict_pipe.Reader.t
+      ; snark_pool_diffs: Snark_pool.Diff.t Strict_pipe.Reader.t }
+  end
+
+  module Stats : sig
+    type t = (* omitted *)
+  end
+
+  module Abstract_peer_graph =
+    Graph.Persistent.Digraph.ConcreteBidirectional (struct
+      type t =
+        | AnyPeer
+        | SpecificPeer of
+      [@@deriving equal, hash]
+    end)
+
+  type t
+
+  (* We can construct the gossip network subsystem using the configuration,
+   * the set of RPC implementations, a state which is shared with the
+   * handlers of the provided RPC implementations. Once constructed, the
+   * gossip network handle will be returned along with pipes for reading
+   * incoming gossip network messages. *)
+  val create :
+       Config.t
+    -> ('state Rpc_intf.t_with_implementation) list
+    -> 'state
+    -> (t * Gossip_pipes.t) Deferred.Or_error.t
+
+  val stats : t -> Stats.t
+
+  (* Query operations now have the ability to express additional validation
+   * logic on a per-request basis, in addition to RPC-wide validation logic
+   * that is defined  *)
+  val query_peer :
+       t
+    -> Peer.t
+    -> ('q, 'r) Rpc_intf.rpc
+    -> 'q
+    -> ?f:('r Envelope.Incoming.t Deferred.t -> Validation_status.t Deferred.t)
+    -> 'r Envelope.Incoming.t Deferred.Or_error.t
+
+  val query_peers :
+       t
+    -> Abstract_peer_graph.t
+    -> ('q, 'r) Rpc_intf.rpc
+    -> 'q
+    -> ?f:('r Envelope.Incoming.t Deferred.t -> Validation_status.t Deferred.t)
+    -> 'r Envelope.Incoming.t Deferred.Or_error.t
+
+  val broadcast : t -> Gossip_message.t -> unit Deferred.t
+
+  val ban_peer : t -> Peer.t -> unit Deferred.t
+end
+```
+
+### Go Implementation
+
+The Go implementation will be fairly similar to how it's structured today. The scope of the state it maintains is more or less the same, the biggest changes introduced in this RFC effect the OCaml code more.  The main work in Go will just be switching it to use Cap'N Proto and use the new message format instead of the old one.
+
+One notable change that can be made is that, since we are moving to a push-based model for libp2p helper metrics, we no longer need to host a prometheus server from Go. However, we will still want the ability to optionally host an http server that exposes the [pprof](https://golang.org/pkg/net/http/pprof/) debugger interface, which we currently support in the metrics server we run.
+
+## Drawbacks
+[drawbacks]: #drawbacks
+
+- a refactor of this scope will take some time to test (given historical context for libp2p work, this could be significant)
+  - COUNTER: we will have to do something like this eventually anyway, better to do it now than later
+
+## Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+- instead of refactoring our current go process integration, we could replace our go helper with a rust process now that there is better libp2p support in rust
```

Comment:
we were discussing this a bit internally also, I believe with rust it would be possible for OCaml and rust to interact directly over FFI so there wouldn't be a need for pipes, which could potentially be faster. however the obvious downside is this would require a full rewrite of the libp2p code, and more changes to the OCaml side as well

Reply:
I think this is something we can (and should) revisit in a year or two when the Rust libp2p Mina work is more mature.